pkg/clustermesh/common/clustermesh.go

```diff
@@ -182,13 +208,44 @@ func (cm *clusterMesh) add(name, path string) {
 }
 
 func (cm *clusterMesh) remove(name string) {
+	const removed = ""
+
 	cm.mutex.Lock()
-	if cluster, ok := cm.clusters[name]; ok {
-		cluster.onRemove()
-		delete(cm.clusters, name)
-		cm.conf.Metrics.TotalRemoteClusters.WithLabelValues(cm.conf.ClusterInfo.Name, cm.conf.NodeName).Set(float64(len(cm.clusters)))
+	defer cm.mutex.Unlock()
+
+	cluster, ok := cm.clusters[name]
+	if !ok {
+		if _, alreadyRemoving := cm.tombstones[name]; alreadyRemoving {
+			// Reset possibly queued add events
+			cm.tombstones[name] = removed
+		}
+
+		return
 	}
-	cm.mutex.Unlock()
+
+	cm.tombstones[name] = removed
+	delete(cm.clusters, name)
+	cm.conf.Metrics.TotalRemoteClusters.WithLabelValues(cm.conf.ClusterInfo.Name, cm.conf.NodeName).Set(float64(len(cm.clusters)))
+
+	cm.wg.Add(1)
+	go func() {
```

Comment:
I'm not sure if doing this async doesn't make it worse for the case when someone changes cluster-name, without changing ClusterID, no?
Previous cluster is going to be deleting while config with a new cluster name will fail on ClusterID reservation?

Reply:
I don't think there would be any difference, as the inotify watcher is non deterministic (as in, it may be, but it depends on the actual file names), and we could either first observe the addition or deletion event.

Reply:
makes sense, thanks for the explanation.